hhefesto/pattern match -- ready to be checked for merge

hhefesto opened the pull request.

hhefesto committed:
Add int and var pattern matching

sfultong commented on src/Telomare/Parser.hs:
```diff
@@ -93,6 +101,81 @@ type TelomareParser = Parsec Void String
+    CaseUP x ls ->
+      -- let dirs = (first ($ x)) <$> (first findInts <$> ls)
+      let dirsOnX :: [UnprocessedParsedTerm]
+          dirsOnX = fst <$> (first ListUP <$> (first (($ x) <$>) <$> (first findInts <$> ls)))
```
this line is hard to read

sfultong commented:
Looks like a good start!

hhefesto committed:
Add String pattern match in case expressions

hhefesto committed:
Fix order for pattern leaves finding

hhefesto committed:
Create test suite for Case expressions

Some cleanup too

sfultong commented on src/Telomare/Resolver.hs:
```diff
@@ -61,6 +59,170 @@ i2c x = TLam (Closed "f") (TLam (Open "x") (inner x))
+        (AppUP (VarUP "abort") $ StringUP "Non-exhaustive patterns in case")
+        (IntUP 0)
+case2annidatedIfs x (aPattern:as) (dirs2IntsOnUPT:bs) (dirs2IntsOnPattern:cs) (dirs2StringOnUPT:ds) (dirs2StringOnPattern:es) (resultAlternative:fs) =
+  let patternVarsOnUPT :: Map String UnprocessedParsedTerm
```
Unused definition?

hhefesto replied:
I wonder why my hls or hlint didn't catch this. Fixed, thanks!

sfultong commented on src/Telomare/Resolver.hs:
```diff
@@ -61,6 +59,170 @@ i2c x = TLam (Closed "f") (TLam (Open "x") (inner x))
+                         (IntUP 1))
+                  (ListUP [ AppUP (AppUP (VarUP "listEqual") dirs2IntsOnUPT) dirs2IntsOnPattern
+                          , AppUP (AppUP (VarUP "listEqual") dirs2StringOnUPT) dirs2StringOnPattern
+                          , pairStructureCheck aPattern x
```
This seems like it will generate some overhead for performing tests for int/string matches where there aren't ints/strings. This wouldn't be a big deal, if Telomare wasn't so slow at the moment :( (I'm making progress on that, I swear!)

hhefesto replied:
Implemented a version with less overhead. let me know what you think

sfultong commented on test/CaseTests.hs:
```diff
@@ -0,0 +1,149 @@
+runCaseExpWithPattern :: (Pattern -> String) -> Pattern -> IO String
+runCaseExpWithPattern p2s p = runTelomareStr $ p2s p
+
+qcPropsCase = testGroup "Property tests on case expressions (QuickCheck)"
```
like to see the quickcheck tests!

hhefesto committed:
Do review changes:

- less overhead on `case2annidatedIfs`
- remove unused `patternVarsOnUPT` binding
